server/src/main/java/org/candlepin/controller/ProductManager.java

```diff
@@ -262,6 +262,69 @@ public Product updateProduct(ProductDTO update, Owner owner, boolean regenerateE
         return updated;
     }
 
+    /**
+     * Pre process the product and provided product hierarchy to build a flat map of products,
+     * which to be processed further.
+     * NOTE - Currently support one level of product & provided product hierarchy.
+     *
+     * @param owner
+     *  The owner for which to import the given product
+     *
+     * @param productData
+     *  A mapping of Red Hat product ID to product DTOs to import
+     *
+     * @param importedContent
+     *  A mapping of Red Hat content ID to content instances to use to lookup and resolve content
+     *  references on the provided product DTOs.
+     *
+     * @return
+     *  A mapping of Red Hat content ID to content entities representing the imported content
+     */
+    @Transactional
+    @Traceable
+    public ImportResult<Product> importProducts(@TraceableParam("owner") Owner owner,
```

Comment:
What exactly are we achieving with this? the original `productData` map that enters this method is already a flat map (contains marketing and provided products equally, as created by the `ImportedEntityCompiler`). This method is simply splitting them up into two maps, and calls the same method for both of them, then merging them together again. So we do the same processing for both maps, with the behaviour unchanged. Is the PR a WIP to get N-tier working?

Reply:
As discussed N-Tier support for multi-level hierarchy is being scope creep at this point, which need to be taken care of
in separate task.
Currently in this PR we are supporting one level of product hierarchy (ie. products having provided products).
In this PR I am splitting the incoming product Map into two maps of products having provided products and product
with no provided products (`productsWithNoProvidedProducts` & `productsWithProvidedProducts`). These two maps are then processed separately, This is necessary to make sure that the child product (eng product) are persisted into DB (regardless of ordering in the input map) before the main product (mkt product).

Reply:
Ah, I see, so this is a preemptive change to make sure the order is correct (eng. products first, then marketing), which otherwise could have caused issues?

Reply:
Yes, the ordering of the products inside `productData` is unpredictable, this change make sure the eng. products are processed and persisted in DB before the marketing products.

Reply:
There is another solution where we set the dummy provided product, in product Update/Create loops. This dummy product is then replaced with original provided products which can be found in one of three lists (ie created, updated, skipped lists) at the time of staging entities. 

This allow us to process `productData` as it is (without worrying about the levels of hierarchy), 
also doesn't require any hibernate cascade to persist the nested entities. This can be easily extended to support N-tier in future where we can process the staged entities in order of depth (bottom to top).